controlplane/kubeadm/internal/controllers/remediation.go

```diff
@@ -107,81 +116,92 @@ func (r *KubeadmControlPlaneReconciler) reconcileUnhealthyMachines(ctx context.C
 
 	// Before starting remediation, run preflight checks in order to verify it is safe to remediate.
 	// If any of the following checks fails, we'll surface the reason in the MachineOwnerRemediated condition.
-	log.WithValues("Machine", klog.KObj(machineToBeRemediated))
-	desiredReplicas := int(*controlPlane.KCP.Spec.Replicas)
-
-	// The cluster MUST have more than one replica, because this is the smallest cluster size that allows any etcd failure tolerance.
-	if controlPlane.Machines.Len() <= 1 {
-		log.Info("A control plane machine needs remediation, but the number of current replicas is less or equal to 1. Skipping remediation", "Replicas", controlPlane.Machines.Len())
-		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP can't remediate if current replicas are less or equal to 1")
-		return ctrl.Result{}, nil
-	}
-
-	// The number of replicas MUST be equal to or greater than the desired replicas. This rule ensures that when the cluster
-	// is missing replicas, we skip remediation and instead perform regular scale up/rollout operations first.
-	if controlPlane.Machines.Len() < desiredReplicas {
-		log.Info("A control plane machine needs remediation, but the current number of replicas is lower that expected. Skipping remediation", "Replicas", desiredReplicas, "CurrentReplicas", controlPlane.Machines.Len())
-		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP waiting for having at least %d control plane machines before triggering remediation", desiredReplicas)
-		return ctrl.Result{}, nil
-	}
+	log.WithValues("Machine", klog.KObj(machineToBeRemediated), "Initialized", controlPlane.KCP.Status.Initialized)
 
-	// The cluster MUST have no machines with a deletion timestamp. This rule prevents KCP taking actions while the cluster is in a transitional state.
-	if controlPlane.HasDeletingMachine() {
-		log.Info("A control plane machine needs remediation, but there are other control-plane machines being deleted. Skipping remediation")
-		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.WaitingForRemediationReason, clusterv1.ConditionSeverityWarning, "KCP waiting for control plane machine deletion to complete before triggering remediation")
+	// Check if KCP is allowed to remediate considering retry limits:
+	// - Remediation cannot happen because retryPeriod is not yet expired.
+	// - KCP already reached MaxRetries limit.
+	hasRetries, retryCount := r.checkRetryLimits(log, machineToBeRemediated, controlPlane)
```

Comment:
What is the meaning of hasRetries, essentially retryAllowed?

Reply:
renamed to canRemediate